### PR TITLE
GitHub actions

### LINKED ISSUE #80: Move from Azure Pipelines to GitHub Actions
I'm currently using Azure Pipelines on my personal Azure account to setup automatic unstable builds/publish and the regular stable build/publish, as well as to deploy the website (via hugo). I should switch this to GitHub Actions (or at least investigate it!).

### LINKED ISSUE #151: Portable build?
It would be great if portable builds were provided as I would like to use this on a computer without admin privileges or on a USB.

### LINKED ISSUE #153: Automatic Update issues
So workspacer should have automatic updates which is provided through the `AutoUpdater.NET` `NuGet` package. However I'm not sure if that is working as intended. 

The following is implemented in `\workspacer\src\workspacer\Program.cs`:

```csharp
#if BRANCH_unstable
            var branch = "unstable";
#elif BRANCH_stable 
            var branch = "stable";
#else
            string branch = null;
#endif
            if (branch != null)
            {
                var xmlUrl = $"https://workspacer.blob.core.windows.net/installers/{branch}.xml";

                AutoUpdater.ApplicationExitEvent += QuitForUpdate;

                System.Timers.Timer timer = new System.Timers.Timer(1000 * 60 * 60);
                timer.Elapsed += (s, e) =>
                {
                    AutoUpdater.Start(xmlUrl);
                };
                timer.Enabled = true;
                AutoUpdater.Start(xmlUrl);
            }
```

Maybe I didn't look carefully enough but I couldn't find a preprocessor directive for either `BRANCH_unstable` or `BRANCH_stable` meaning the check for updates never happens because `branch`will always evaluate to `null`. I tested this by editing the Assembly Version and the update does trigger correctly when the directives are present. I still don't quite understand the purpose of the 1hr timer though...

A second potential issue is that the user has no way of configuring this behaviour without going into the source files. I think it'd be nice to have this accessible through the config file instead. 

A a side note would implementing checksums for the downloads be possible? Would be nice if there's a way verify the integrity of the download prior to execution.

Happy to write something up for this if there's interest to have this feature.

### PR DESCRIPTION
**Pull request by sitiom**
Github Actions.

Would resolve #80, resolve #151, resolve #153.

`tests.yml` would build and test in PRs to master, `unstable.yml` would build and prerelease to the `unstable` tag every push to master (would overwrite the previous unstable), and `stable.yml` would build and release each tag push. This works on my fork at least.

Of course, Autoupdater.NET has also been made to track stable and unstable releases on GitHub as well.

Edit: Added site deploy now

### CODE REVIEW

**Comment by josteink** [comment]
Holy moly. That's a nice PR! ❤️

File: .github/workflows/stable.yml
Comment (josteink) [inline .github/workflows/stable.yml:18]: Since this is .Net Core can't/shouldn't we AOT-compile this using `dotnet publish` instead?

File: .github/workflows/tests.yml
Comment (josteink) [inline .github/workflows/tests.yml:24]: Any reason to prefer xUnit-cli over plain `dotnet test`?

**Review by josteink — APPROVED** [review]
Looks mostly good, but I have a few small questions/suggestions. 

Nothing critical or blocking a merge from my point of view.

File: .github/workflows/stable.yml
Comment (sitiom) [reply .github/workflows/stable.yml:18]: For some reason, workspacer fails to run on `dotnet publish` for me.

Comment (josteink) [reply .github/workflows/stable.yml:18]: Fair enough.

File: .github/workflows/tests.yml
Comment (sitiom) [reply .github/workflows/tests.yml:24]: A console runner is needed and xunit-cli wraps around it.

Comment (josteink) [reply .github/workflows/tests.yml:24]: Not sure I follow. I use dotnet test with xUnit in my CI jobs without any additional setup. 

Care to elaborate a little further?

Comment (sitiom) [reply .github/workflows/tests.yml:24]: Oh, for some reason, tests get skipped when I used  `dotnet test` 🤔.
```
  Determining projects to restore...
  All projects are up-to-date for restore.
  workspacer.Shared -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Shared\bin\Debug\net5.0-windows\win10-x64\workspacer.Shared.dll
  workspacer.FocusIndicator -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.FocusIndicator\bin\Debug\net5.0-windows\win10-x64\workspacer.FocusIndicator.dll
  workspacer.Native -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Native\bin\Debug\net5.0-windows\win10-x64\workspacer.Native.dll
  workspacer.Bar -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Bar\bin\Debug\net5.0-windows\win10-x64\workspacer.Bar.dll
  workspacer.Shared.Tests -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Shared.Tests\bin\Debug\net5.0-windows\win10-x64\workspacer.Shared.Tests.dll
  workspacer.ActionMenu -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.ActionMenu\bin\Debug\net5.0-windows\win10-x64\workspacer.ActionMenu.dll
Test run for C:\Users\sitiom\source\repos\workspacer\src\workspacer.Shared.Tests\bin\Debug\net5.0-windows\win10-x64\workspacer.Shared.Tests.dll (.NETCoreApp,Version=v5.0)
Microsoft (R) Test Execution Command Line Tool Version 16.8.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
  workspacer.Watcher -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Watcher\bin\Debug\net5.0-windows\win10-x64\workspacer.Watcher.dll
A total of 1 test files matched the specified pattern.
  workspacer -> C:\Users\sitiom\source\repos\workspacer\src\workspacer\bin\Debug\net5.0-windows\win10-x64\workspacer.dll
  workspacer.Tests -> C:\Users\sitiom\source\repos\workspacer\src\workspacer.Tests\bin\Debug\net5.0-windows\win10-x64\workspacer.Tests.dll
Test run for C:\Users\sitiom\source\repos\workspacer\src\workspacer.Tests\bin\Debug\net5.0-windows\win10-x64\workspacer.Tests.dll (.NETCoreApp,Version=v5.0)
Microsoft (R) Test Execution Command Line Tool Version 16.8.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     3, Skipped:     0, Total:     3, Duration: 4 ms - workspacer.Shared.Tests.dll (net5.0)

Passed!  - Failed:     0, Passed:     7, Skipped:     0, Total:     7, Duration: 7 ms - workspacer.Tests.dll (net5.0)
```

and xunit-cli fixed it, so I used it:

```
xUnit.net Console Runner v2.4.1 (64-bit .NET 5.0.0)
  Discovering: workspacer.Shared.Tests
  Discovered:  workspacer.Shared.Tests
  Starting:    workspacer.Shared.Tests
    workspacer.Shared.Tests.ColorTests.Creating_Color_With_Invalid_RGB_Components_Throws [STARTING]
    workspacer.Shared.Tests.ColorTests.Creating_Color_With_Invalid_RGB_Components_Throws [FINISHED] Time: 0.0034181s
    workspacer.Shared.Tests.ColorTests.Creating_Color_With_Valid_RGB_Components_Succeeds [STARTING]
    workspacer.Shared.Tests.ColorTests.Creating_Color_With_Valid_RGB_Components_Succeeds [FINISHED] Time: 0.0000679s
    workspacer.Shared.Tests.ColorTests.Different_Colors_Does_Not_Have_Hash_Collision [STARTING]
    workspacer.Shared.Tests.ColorTests.Different_Colors_Does_Not_Have_Hash_Collision [FINISHED] Time: 0.0015202s
  Finished:    workspacer.Shared.Tests
  Discovering: workspacer.Tests
  Discovered:  workspacer.Tests
  Starting:    workspacer.Tests
    workspacer.Tests.SubTests.Does_Not_Have_Hash_Collisions_Between_Different_Keys [STARTING]
    workspacer.Tests.SubTests.Does_Not_Have_Hash_Collisions_Between_Different_Keys [FINISHED] Time: 0.0010873s
    workspacer.Tests.SubTests.LeftCtrl_Does_Not_Equal_RightControl [STARTING]
    workspacer.Tests.SubTests.LeftCtrl_Does_Not_Equal_RightControl [FINISHED] Time: 0.0040807s
    workspacer.Tests.SubTests.LeftCtrl_RightCtrl_And_Ctrl_Has_Same_HashCode [STARTING]
    workspacer.Tests.SubTests.LeftCtrl_RightCtrl_And_Ctrl_Has_Same_HashCode [FINISHED] Time: 0.0002138s
    workspacer.Tests.SubTests.RightAlt_Does_Not_Equal_LeftAlt [STARTING]
    workspacer.Tests.SubTests.RightAlt_Does_Not_Equal_LeftAlt [FINISHED] Time: 0.0000521s
    workspacer.Tests.SubTests.RightAlt_LeftAlt_And_Alt_Has_Same_HashCode [STARTING]
    workspacer.Tests.SubTests.RightAlt_LeftAlt_And_Alt_Has_Same_HashCode [FINISHED] Time: 0.0000338s
    workspacer.Tests.SubTests.RightWin_Does_Not_Equal_LeftWin [STARTING]
    workspacer.Tests.SubTests.RightWin_Does_Not_Equal_LeftWin [FINISHED] Time: 0.0000414s
    workspacer.Tests.SubTests.RightWin_LeftWin_And_WinHas_Same_HashCode [STARTING]
    workspacer.Tests.SubTests.RightWin_LeftWin_And_WinHas_Same_HashCode [FINISHED] Time: 0.00004s
  Finished:    workspacer.Tests
=== TEST EXECUTION SUMMARY ===
   workspacer.Shared.Tests  Total:  3, Errors: 0, Failed: 0, Skipped: 0, Time: 0.068s
   workspacer.Tests         Total:  7, Errors: 0, Failed: 0, Skipped: 0, Time: 0.008s
                                   --          -          -           -        ------
                      GRAND TOTAL: 10          0          0           0        0.076s (0.148s)
```

Comment (josteink) [reply .github/workflows/tests.yml:24]: If so, that's something which IMO needs to be fixed too. 

How are you invoking dotnet test? Against main sln-file or something else?

Comment (sitiom) [reply .github/workflows/tests.yml:24]: > If so, that's something which IMO needs to be fixed too.
> 
> How are you invoking dotnet test? Against main sln-file or something else?

On the sln file.

Comment (sitiom) [reply .github/workflows/tests.yml:24]: Oh actually nevermind, I didn't read it clearly and it actually passed according to the output above. Thanks for pointing this out.

**Comment by josteink** [comment]
LGTM!

**Comment by josteink** [comment]
To be clear I think this looks good, but for obvious reasons I'm going to hold off merging this until @rickbutton approves 😅

Also: might be nice with a build-badge in the README too?

File: .github/workflows/stable.yml
Comment (sitiom) [reply .github/workflows/stable.yml]: @josteink It turns out the plugins weren't copied on publish and only on build. Fixed it now.

**Comment by josteink** [comment]
This PR is getting pretty big. 

Can't the last commit be proposed as it's own PR?

I don't see how it's relevant to GitHub actions.

**Comment by sitiom** [comment]
> This PR is getting pretty big.
> 
> Can't the last commit be proposed as it's own PR?
> 
> I don't see how it's relevant to GitHub actions.

It's part of getting the branch overridden by the user on the config. Perhaps I should squash the commits to reduce it.

File: src/workspacer/workspacer.csproj
Comment (josteink) [inline src/workspacer/workspacer.csproj:21]: Target**s** is plural. I believe you should be able to use the same target-block for both Build and Publish to avoid code-duplication.

https://docs.microsoft.com/en-us/visualstudio/msbuild/target-element-msbuild?view=vs-2019

Comment (sitiom) [reply src/workspacer/workspacer.csproj:21]: According to the docs, it would run "after the specified target or targets", which means it would only run when both build and publish runs. Also, the directory of the build and publish folders are different ( `TargetDir` vs `PublishDir`).

Comment (josteink) [reply src/workspacer/workspacer.csproj:21]: You're right. I missed that. 

Let's keep it as is then.

File: .github/workflows/stable.yml
Comment (rickbutton) [inline .github/workflows/stable.yml:14]: love it

File: README.md
Comment (rickbutton) [inline README.md:9]:
```suggestion
[check out the quick start guide](https://workspacer.org/quickstart/)
```

File: src/workspacer/Program.cs
Comment (rickbutton) [inline src/workspacer/Program.cs:88]:
```suggestion
                ChangelogURL = isStable ? "https://workspacer.org/changelog/" : "https://github.com/rickbutton/workspacer/releases/unstable",
```

**Review by rickbutton — CHANGES_REQUESTED** [review]
this is unbelievably awesome. thank you for this, this was clearly a ton of work, I really appreciate it!

I requested just a couple of tiny changes to switch the domain back to workspacer.org, which is CNAME-d to rickbutton.github.io anyway.

In terms of transitioning to this new pipeline, once this is merged and working, we can deploy a version bump, and then modifiy the `releases.xml` file in the old Azure blob storage to point to the new GitHub release.

**Review by rickbutton — APPROVED** [review]

**Pull request merged**
